cpp/diffraction/StaticStructureFactorDebye.cc

```diff
@@ -0,0 +1,104 @@
+    \brief Routines for computing static structure factors.
+*/
+
+namespace freud { namespace diffraction {
+
+StaticStructureFactorDebye::StaticStructureFactorDebye(unsigned int bins, float k_max, float k_min)
+{
+    if (bins == 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDebye requires a nonzero number of bins.");
+    }
+    if (k_max <= 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDebye requires k_max to be positive.");
+    }
+    if (k_max <= k_min)
+    {
+        throw std::invalid_argument("StaticStructureFactorDebye requires that k_max must be greater than k_min.");
+    }
+
+    // Construct the Histogram object that will be used to track the structure factor
+    auto axes
+        = StaticStructureFactorDebyeHistogram::Axes {std::make_shared<util::RegularAxis>(bins, k_min, k_max)};
+    m_histogram = StaticStructureFactorDebyeHistogram(axes);
+    m_local_histograms = StaticStructureFactorDebyeHistogram::ThreadLocalHistogram(m_histogram);
+    m_min_valid_k = std::numeric_limits<float>::infinity();
+    m_structure_factor.prepare(bins);
+}
+
+void StaticStructureFactorDebye::accumulate(const freud::locality::NeighborQuery* neighbor_query,
+                                             const vec3<float>* query_points, unsigned int n_query_points, unsigned int n_total)
+{
+    auto const& box = neighbor_query->getBox();
+
+    // The r_max should be just less than half of the smallest side length of the box
+    auto const box_L = box.getL();
+    auto const min_box_length
+        = box.is2D() ? std::min(box_L.x, box_L.y) : std::min(box_L.x, std::min(box_L.y, box_L.z));
+    auto const r_max = std::nextafter(float(0.5) * min_box_length, float(0));
+    auto const points= neighbor_query->getPoints();
+    auto const n_points= neighbor_query->getNPoints();
+
+    // The minimum k value of validity is 4 * pi / L, where L is the smallest side length.
+    // This is equal to 2 * pi / r_max.
+    m_min_valid_k = std::min(m_min_valid_k, freud::constants::TWO_PI / r_max);
+
+    auto distances = std::vector<float>(n_points * n_query_points);
+    box.computeAllDistances(points, n_points, query_points, n_query_points, distances.data());
+
+    auto const k_bin_centers = m_histogram.getBinCenters()[0];
+
+    util::forLoopWrapper(0, m_histogram.getAxisSizes()[0], [&](size_t begin_k_index, size_t end_k_index) {
+        for (size_t k_index = begin_k_index; k_index < end_k_index; ++k_index)
+        {
+            auto const k = k_bin_centers[k_index];
+            double S_k = 0.0;
+            for (size_t distance_index = 0; distance_index < distances.size(); ++distance_index)
+            {
+                auto const distance = distances[distance_index];
+                S_k += util::sinc(k * distance);
+            }
+            S_k /= static_cast<double>(n_total);
+            m_local_histograms.increment(k_index, S_k);
+        };
+    });
+    m_frame_counter++;
+    m_reduce = true;
+}
```

Comment:
There's a lot of `auto` variables here. I think some of them are fine to be left as `auto`, but many like in lines 56, 57, and 72 should be declared as their actual type to prevent confusion.

Reply:
I've changed some auto variables. Let me know if you're happy with changes.

Reply:
@tommy-waltmann @DomFijan Many C++ experts consider it good style to adopt "Almost Always Auto." References:
- https://www.fluentcpp.com/2020/12/04/auto-for-types-but-not-for-concepts/
- https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

I see no advantage to dropping `auto` in the cases @tommy-waltmann pointed out.

Reply:
Okay, I'll buy the arguments from the articles. We can revert those lines.

Reply:
The commit was reverted.